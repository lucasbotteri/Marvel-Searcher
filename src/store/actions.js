import {
  SET_LOADING_CHARACTERS_LIST,
  SET_CHARACTERS_LIST,
  SET_CHARACTER_SELECTED,
  HIDE_MODAL,
  SET_COMICS,
  SET_LOADING_COMICS,
  SET_COMIC,
  SET_LOADING_COMIC,
  SAVE_FAV,
  SHOW_FAVS,
  REMOVE_FAV,
  HIDE_FAVS,
} from "./constants";
import {
  getCharacterByName,
  getRandomCharacter,
  getCharacterComics,
  getComic,
} from "../service";

export const setLoadingCharacters = () => ({
  type: SET_LOADING_CHARACTERS_LIST,
});

export const setCharacters = (characters) => ({
  type: SET_CHARACTERS_LIST,
  payload: characters,
});

export const setCharacterSelected = (id, name) => ({
  type: SET_CHARACTER_SELECTED,
  payload: { id, name },
});

export const hideModal = () => ({
  type: HIDE_MODAL,
});

export const setComics = (comics) => ({
  type: SET_COMICS,
  payload: comics,
});

export const setLoadingComics = () => ({
  type: SET_LOADING_COMICS,
});

export const setComic = (comics) => ({
  type: SET_COMIC,
  payload: comics,
});

export const setLoadingComic = () => ({
  type: SET_LOADING_COMIC,
});

export const saveFav = (character) => ({
  type: SAVE_FAV,
  payload: character,
});

export const showFavs = () => ({
  type: SHOW_FAVS,
});

export const hideFavs = () => ({
  type: HIDE_FAVS,
});

export const removeFav = (id) => ({
  type: REMOVE_FAV,
  payload: { id },
});

export const searchCharacter = async (dispatch, characterName) => {
  dispatch(setLoadingCharacters());
  const characters = await getCharacterByName(characterName);
  dispatch(setCharacters(characters));
};

export const searchRandomCharacter = async (dispatch) => {
  dispatch(setLoadingCharacters());
  const characters = await getRandomCharacter();
  dispatch(setCharacters(characters));
};

export const searchCharacterComics = async (
  dispatch,
  characterId,
  comicFilter
) => {
  dispatch(setLoadingComics());
  const comics = await getCharacterComics(characterId, comicFilter);
  dispatch(setComics(comics));
};

export const searchComic = async (dispatch, comicId) => {
  dispatch(setLoadingComic());
  const comic = await getComic(comicId);
  dispatch(setComic(comic));
};
