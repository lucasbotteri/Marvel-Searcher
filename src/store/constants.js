export const LOADING_CHARACTERS_LIST = "LOADING_CHARACTERS_LIST";
export const SET_CHARACTERS_LIST = "SET_CHARACTERS_LIST";
export const SET_CHARACTER_SELECTED = "SET_CHARACTER_SELECTED";
export const HIDE_MODAL = "HIDE_MODAL";
export const SET_COMICS = "SET_COMICS";
export const SET_LOADING_COMICS = "SET_LOADING_COMICS";
