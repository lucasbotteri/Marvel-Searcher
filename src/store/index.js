export { default as StoreProvider, StoreContext } from "./StoreProvider";
