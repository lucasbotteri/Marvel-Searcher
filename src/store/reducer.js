import {
  LOADING_CHARACTERS_LIST,
  SET_CHARACTERS_LIST,
  SET_CHARACTER_SELECTED,
  HIDE_MODAL,
} from "./constants";

const reducer = (state, action) => {
  switch (action.type) {
    case LOADING_CHARACTERS_LIST:
      return { ...state, characters: [], isCharacterListLoading: true };
    case SET_CHARACTERS_LIST:
      return {
        ...state,
        characters: action.payload,
        isCharacterListLoading: false,
        isShowingModal: false,
      };
    case SET_CHARACTER_SELECTED:
      return {
        ...state,
        characterSelected: action.payload,
        isShowingModal: true,
      };
    case HIDE_MODAL:
      return {
        ...state,
        isShowingModal: false,
      };
    default:
      throw new Error("Invalid Action");
  }
};

export default reducer;
