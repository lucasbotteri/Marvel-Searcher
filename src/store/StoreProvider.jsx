import React, { useReducer } from "react";
import PropTypes from "prop-types";
import reducer from "./reducer";
import * as actions from "./actions";

const INITAL_STATE = {
  characters: [],
  loadingCharacters: false,
  fetchedCharactersSuccessful: true,
};

export const StoreContext = React.createContext();

const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, INITAL_STATE);
  const value = {
    ...state,
    fetchCharacters: (data) => actions.fetchCharacters(dispatch, data),
  };
  return (
    <StoreContext.Provider value={value}>{children}</StoreContext.Provider>
  );
};

StoreProvider.propTypes = {
  children: PropTypes.element.isRequired,
};

export default StoreProvider;
