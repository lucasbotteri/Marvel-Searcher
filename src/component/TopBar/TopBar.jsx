import React, { useContext } from "react";
import styled from "styled-components";
import Star from "../Star";
import SearchBox from "../SearchBox";
import Logo from "../../logo.svg";
import { StoreContext } from "../../store";

const FixedTopBarWrapper = styled.nav`
  align-items: center;
  background-color: ${(props) => props.theme.primaryColor};
  box-shadow: 0 1px 4px 0 ${(props) => props.theme.secondaryColor};
  display: flex;
  height: 50px;
  justify-content: space-around;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 9999;
`;

const TopBarFav = styled(Star)`
  color: ${(props) => props.theme.secondaryColor};
`;

const VerticalDivider = styled.div`
  border-left: 1px solid ${(props) => props.theme.secondaryColor};
  height: 75%;
`;

const MarvelLogo = styled.img`
  border-left: 1px solid ${(props) => props.theme.secondaryColor};
  height: 66%;
`;

const TopBar = () => {
  const { searchCharacter } = useContext(StoreContext);
  const onSearch = (searchText) => {
    searchCharacter(searchText);
  };

  return (
    <FixedTopBarWrapper>
      <MarvelLogo src={Logo} alt="marvel logo" />
      <VerticalDivider />
      <SearchBox onSearch={onSearch} />
      <TopBarFav isSolid={false} />
      <VerticalDivider />
    </FixedTopBarWrapper>
  );
};

export default TopBar;
