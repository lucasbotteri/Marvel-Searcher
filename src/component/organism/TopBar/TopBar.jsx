import React, { useContext } from "react";
import styled from "styled-components";
import Star from "../../molecule/Star";
import SearchBox from "../../molecule/SearchBox";
import Logo from "../../../logo.svg";
import { StoreContext } from "../../../store";

const MARVEL_COMIC_URL_REGEX = /https:\/\/www\.marvel\.com\/comics\/issue\/[0-9]+\//;
const MARVEL_COMIC_ID_REGEX = /\/[0-9]+\//;

const FixedTopBarWrapper = styled.nav`
  align-items: center;
  background-color: ${(props) => props.theme.primaryColor};
  box-shadow: 0 1px 4px 0 ${(props) => props.theme.secondaryColor};
  display: flex;
  height: 50px;
  justify-content: space-around;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 9999;
`;

const TopBarFav = styled(Star)`
  color: ${(props) => props.theme.secondaryColor};
`;

const VerticalDivider = styled.div`
  border-left: 1px solid ${(props) => props.theme.secondaryColor};
  height: 75%;
`;

const MarvelLogo = styled.img`
  border-left: 1px solid ${(props) => props.theme.secondaryColor};
  height: 66%;
`;

const TopBar = () => {
  const {
    searchCharacter,
    searchComic,
    showFavs,
    isShowingFavs,
    hideFavs,
  } = useContext(StoreContext);

  const onSearch = (searchText) => {
    if (searchText) {
      const isAComicSearch = MARVEL_COMIC_URL_REGEX.test(searchText);
      if (isAComicSearch) {
        const comicId = searchText.match(MARVEL_COMIC_ID_REGEX)[0];
        const parsedComicId = comicId.slice(1, comicId.length - 1);
        searchComic(parsedComicId);
      } else {
        searchCharacter(searchText);
      }
    }
  };

  return (
    <FixedTopBarWrapper>
      <MarvelLogo src={Logo} alt="marvel logo" />
      <VerticalDivider />
      <SearchBox onSearch={onSearch} />
      <TopBarFav
        isSolid={isShowingFavs}
        onClick={isShowingFavs ? hideFavs : showFavs}
      />
      <VerticalDivider />
    </FixedTopBarWrapper>
  );
};

export default TopBar;
