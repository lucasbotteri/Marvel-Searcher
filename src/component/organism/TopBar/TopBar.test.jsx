import React from "react";
import { mount } from "enzyme";
import TopBar from "./TopBar";
import { StoreContext } from "../../../store";

const characters = [
  {
    id: 1014997,
    name: "Deathstrike (Ultimate)",
    thumbnail: "someurl.com",
  },
];

const contextStore = {
  searchCharacter: () => {},
  searchComic: () => {},
  showFavs: () => {},
  hideFavs: () => {},
  isShowingFavs: false,
};

const mountTopBar = () => {
  return mount(
    <StoreContext.Provider value={contextStore}>
      <TopBar characters={characters} />
    </StoreContext.Provider>
  );
};

describe("TopBar", () => {
  it("Should render without crashing", () => {
    const wrapper = mountTopBar();
    expect(wrapper).toBeTruthy();
  });

  it("Should match snapshot", () => {
    const wrapper = mountTopBar();
    expect(wrapper).toMatchSnapshot();
  });
});
