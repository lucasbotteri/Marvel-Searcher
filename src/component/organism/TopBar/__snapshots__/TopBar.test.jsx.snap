// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar Should match snapshot 1`] = `ReactWrapper {}`;
